ui/src/image_annotator_polygon.ts

```diff
@@ -1,41 +1,113 @@
 import { F, S, U } from "h2o-wave"
-import { DrawnPoint, DrawnShape, ImageAnnotatorPoint, Position } from "./image_annotator"
+import { DrawnPoint, DrawnShape, ImageAnnotatorPoint, ImageAnnotatorRect, Position } from "./image_annotator"
 import { ARC_RADIUS } from "./image_annotator_rect"
 
 export class PolygonAnnotator {
   private currPolygonPoints: ImageAnnotatorPoint[] = []
+  private boundaryRect: ImageAnnotatorRect | undefined
   private draggedPoint: ImageAnnotatorPoint | null = null
   private draggedShape: DrawnShape | null = null
+  private ctx: CanvasRenderingContext2D | null
 
-  constructor(private ctx: CanvasRenderingContext2D) {
-    this.ctx.lineWidth = 2
+  constructor(private canvas: HTMLCanvasElement) {
+    this.ctx = canvas.getContext('2d')
+    if (this.ctx) this.ctx.lineWidth = 2
   }
 
   resetDragging() {
+    // Update the boundaries of the polygon when point dragging ends.
+    const draggedPolygon = this.draggedShape?.shape.polygon
+    if (this.draggedPoint && draggedPolygon) {
+      this.boundaryRect = draggedPolygon.boundaryRect
+      this.updateBoundaryRect(this.draggedPoint)
+      draggedPolygon.boundaryRect = this.boundaryRect
+      this.boundaryRect = undefined
+    }
+
     this.draggedPoint = null
     this.draggedShape = null
   }
 
   cancelAnnotating() {
     this.currPolygonPoints = []
+    this.boundaryRect = undefined
+  }
+
+  updateBoundaryRect({ x, y }: { x: U, y: U }) {
+    // TODO: Update boundary when increasing the x1, y1 or decreasing the x2, y2.
+    if (this.boundaryRect) {
+      if (x < this.boundaryRect.x1) this.boundaryRect.x1 = x
+      if (x > this.boundaryRect.x2) this.boundaryRect.x2 = x
+      if (y < this.boundaryRect.y1) this.boundaryRect.y1 = y
+      if (y > this.boundaryRect.y2) this.boundaryRect.y2 = y
+    } else this.boundaryRect = { x1: x, x2: x, y1: y, y2: y }
+  }
+
+  addCurrPolygonPoint({ x, y }: { x: U, y: U }) {
+    this.currPolygonPoints.push({ x, y })
+    this.updateBoundaryRect({ x, y })
+  }
+
+  removeLastPoint() {
+    // TODO: Update boundaries on remove.
+    this.currPolygonPoints.pop()
+  }
+
+  finishPolygon(tag: S) {
+    const
+      { x, y } = this.currPolygonPoints[0],
+      newPolygon = {
+        shape: {
+          polygon: {
+            vertices: [...this.currPolygonPoints],
+            boundaryRect: this.boundaryRect
+          }
+        },
+        tag
+      },
+      isPolygon = this.currPolygonPoints.length > 2
```

Comment:
Shouldn't we check this as the first thing and do an early return if we do have < 2 points?